x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/logging/DeprecationIndexingComponent.java

```diff
@@ -0,0 +1,141 @@
+
+        final LoggerContext context = (LoggerContext) LogManager.getContext(false);
+        final Configuration configuration = context.getConfiguration();
+
+        final EcsLayout ecsLayout = ECSJsonLayout.newBuilder().setType("deprecation").setConfiguration(configuration).build();
+
+        this.filter = new RateLimitingFilter();
+        this.appender = new DeprecationIndexingAppender("deprecation_indexing_appender", filter, ecsLayout, consumer);
+    }
+
+    @Override
+    protected void doStart() {
+        this.appender.start();
+        Loggers.addAppender(LogManager.getLogger("org.elasticsearch.deprecation"), this.appender);
+    }
+
+    @Override
+    protected void doStop() {
+        Loggers.removeAppender(LogManager.getLogger("org.elasticsearch.deprecation"), this.appender);
+        this.appender.stop();
+    }
+
+    @Override
+    protected void doClose() {
+        this.processor.close();
+    }
+
+    /**
+     * Listens for changes to the cluster state, in order to know whether to toggle indexing
+     * and to set the cluster UUID and node ID. These can't be set in the constructor because
+     * the initial cluster state won't be set yet.
+     *
+     * @param event the cluster state event to process
+     */
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        final ClusterState state = event.state();
+        final boolean newEnabled = WRITE_DEPRECATION_LOGS_TO_INDEX.get(state.getMetadata().settings());
+        if (appender.isEnabled() != newEnabled) {
+            // We've flipped from disabled to enabled. Make sure we start with a clean cache of
+            // previously-seen keys, otherwise we won't index anything.
+            if (newEnabled) {
+                this.filter.reset();
+            }
+            appender.setEnabled(newEnabled);
+        }
+    }
+
+    /**
+     * Constructs a bulk processor for writing documents
+     * @param client the client to use
+     * @param settings the settings to use
+     * @return an initialised bulk processor
+     */
+    private BulkProcessor getBulkProcessor(Client client, Settings settings) {
+        final OriginSettingClient originSettingClient = new OriginSettingClient(client, ClientHelper.DEPRECATION_ORIGIN);
+        final BulkProcessor.Listener listener = new DeprecationBulkListener();
+
+        // This configuration disables the size count and size thresholds,
+        // and instead uses a scheduled flush only. This means that calling
+        // processor.add() will not block the calling thread.
+        return BulkProcessor.builder(originSettingClient::bulk, listener)
+            .setBackoffPolicy(BackoffPolicy.exponentialBackoff(TimeValue.timeValueMillis(1000), 3))
+            .setConcurrentRequests(Math.max(2, EsExecutors.allocatedProcessors(settings)))
+            .setBulkActions(-1)
+            .setBulkSize(new ByteSizeValue(-1, ByteSizeUnit.BYTES))
```

Comment:
we actually want the bulk actions and bulk size to be large values with a reasonably short flush interval. setting them to -1 will cause the execution to happen one at a time (removing the bulk from the bulk processor) and more likely to happen on the calling thread. setting to large values allows the bulks to build up until the flush takes over and submits the request. 

Also, it even with flush it does block a little bit while mutating the bulk request, but that should be a very fast operation.

Reply:
I set them to -1 after reading the source of `BulkProcessor`, and the only place I found where `bulkActions` and `bulkSize` are used is in [isOverTheLimit()](https://github.com/elastic/elasticsearch/blob/f72b893fd313810c4d6bf98585d64bc41958b976/server/src/main/java/org/elasticsearch/action/bulk/BulkProcessor.java#L452), and in both cases, if the value is -1 then the check is skipped. That implies that the bulk request can grow in document count or request size indefinitely. What am I missing?

Also, what flush interval are you thinking of? Less than 5s?

Reply:
> What am I missing?

nothing ... it's me, sorry for the noise. (misremembered what that magic number does and a case of confirmation bias spot checking the code) 

5s should be fine